### PR TITLE
Register Datasource Picker in the top nav menu for Get Started Tab

### LINKED ISSUE #1795: [FEATURE] Add dependency to dataSourceManagement bundle
**Is your feature request related to a problem?**
Acceptance criteria: in the opensearch_dashboards.json file, add dataSource as required bundle, eg. "requiredBundles": ["dataSourceManagement"]

**What solution would you like?**
Adding a dependency to the dataSourceManagement bundle and plugin tests/builds pass with and without multiple datasources enabled. 

**What alternatives have you considered?**
None

**Do you have any additional context?**
This is a pre-requisite to consuming the shared/common DataSource Picker component that is shared among plugins/core.
More context can be found here: 
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5717

### LINKED ISSUE #1796: [FEATURE] Support multiple datasources in the "Get Started" tab
**Is your feature request related to a problem?**
Support multiple datasources functionality in the "Get Started" tab. This should be a relatively small task since this tab mainly contains links to other tabs. It does have a "Purge cache" button, which needs to be done in the context of the current datasource.

**Acceptance criteria**
- Purge cache button is correctly done on the selected datasource (do we want to have a modal with extra confirmation that this is happening on XXX cluster)
- Data source picker shared component is present on the page with a sensible UX
- Data source selections that should be persisted across tabs is accurately persisted
- Selected datasource prior to landing on this page is correctly populated (if applicable), and defaults are selected if not applicable

**What solution would you like?**
Support multiple datasources in the "Get Started" tab. 

**What alternatives have you considered?**
None

**Do you have any additional context?**
Depends on #1795 to have data source picker available.

### LINKED ISSUE #1816: [FEATURE] Generalize DataSourcePicker/datasourceid usage
**Is your feature request related to a problem?**
- After this pr is merged: https://github.com/opensearch-project/security-dashboards-plugin/pull/1810, there is some cleanup that is necessary:
- Hook up selected datasource to a global state/context/store that can be read and modified on each tab - investigate options here
- Clean up datasource picker registration (currently it would be globally registered), need to figure out how to register it only in our plugin, be able to change it from writeable to read only on different screens
- Add testing framework/e2e testing framework to verify that correct API calls are being made - maybe blocked on https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4524 - possible workaround is to run with security on ```http://``` instead of ```https://```

### PR DESCRIPTION
**Pull request by derek-ho**
### Description
Registers the new DataSource picker component from core in the top nav menu for the "Get Started" Tab. 

### Category
[Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation]
New feature

### Why these changes are required?
- Resolves #1796
- Resolves #1816
- Resolves #1795 

### What is the old behavior before changes and new behavior after changes?
Old behavior is no selector, new behavior is shown in this video: 

https://github.com/opensearch-project/security-dashboards-plugin/assets/26328171/6554d631-e5ca-461d-a044-23a7556a5b08

### Issues Resolved
[List any issues this PR will resolve (Is this a backport? If so, please add backport PR # and/or commits #)]

Fix: #1796, Fix: #1816 Fix: #1795 

### Testing
Manual testing, unit tests added, e2e cypress test added

### Check List
- [x] New functionality includes testing
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/cypress-test-multidatasources-enabled-e2e.yml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+      - name: Set env
+        run: |
+          opensearch_version=$(node -p "require('./package.json').opensearchDashboards.version")
+          plugin_version=$(node -p "require('./package.json').version")
+          echo "OPENSEARCH_VERSION=$opensearch_version" >> $GITHUB_ENV
+          echo "PLUGIN_VERSION=$plugin_version" >> $GITHUB_ENV
+        shell: bash
+
+      - name: Create remote OpenSearch Config
```
Comment (DarshitChanpura) [inline .github/workflows/cypress-test-multidatasources-enabled-e2e.yml:34]: out of curiosity, why do we need a custom opensearch.yml?

File: .github/workflows/cypress-test-multidatasources-enabled-e2e.yml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        if: ${{ runner.os == 'Linux' }}
+        run: | 
+          cat << 'EOT' > remote_opensearch.yml
```
Comment (derek-ho) [inline .github/workflows/cypress-test-multidatasources-enabled-e2e.yml:37]: This is necessary because currently ssl verification is enforced on the second datasource - thus we need to flip the https setting to false, and change the port number. Once the change in core goes through to configure ssl verification, this can be simplified to just changing the port number.

File: public/apps/configuration/panels/get-started.tsx [hunk 3]
```diff
@@ -158,7 +160,17 @@ const setOfSteps = [
   },
 ];
 
+export function GetClusterDescription(dataSourceEnabled: boolean, cluster: Cluster) {
```
Comment (DarshitChanpura) [inline public/apps/configuration/panels/get-started.tsx:163]: rename it to getClusterLabel?

File: public/apps/configuration/panels/get-started.tsx [hunk 4]
```diff
@@ -158,7 +160,17 @@ const setOfSteps = [
...
+  if (dataSourceEnabled) {
+    return `for ${cluster.label || 'Local cluster'}`;
+  }
+  return '';
+}
+
 export function GetStarted(props: AppDependencies) {
+  const dataSourceEnabled = !!props.securityPluginStartDeps.dataSource?.dataSourceEnabled;
+  const [dataSource, setDataSource] = useState<Cluster>({ id: '', label: '' });
```
Comment (DarshitChanpura) [inline public/apps/configuration/panels/get-started.tsx:172]: can you add a test that this useState exhibits correct behavior?

File: public/apps/configuration/top-nav-menu.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import { PLUGIN_NAME } from '../../../common';
+import { AppDependencies } from '../types';
+import { Cluster } from '../../types';
+
+export interface TopNavMenuProps extends AppDependencies {
+  dataSourcePickerReadOnly: boolean;
+  setDatasourceId: React.Dispatch<React.SetStateAction<Cluster>>;
+}
+
+export const SecurityPluginTopNavMenu = (props: TopNavMenuProps) => {
```
Comment (DarshitChanpura) [inline public/apps/configuration/top-nav-menu.tsx:29]: rename it to `SecureDataSourceSelector`. Thoughts?

File: public/plugin.ts [hunk 6]
```diff
@@ -112,7 +112,14 @@ export class SecurityPlugin
           excludeFromDisabledTransportCategories(config.disabledTransportCategories.exclude);
           excludeFromDisabledRestCategories(config.disabledRestCategories.exclude);
 
-          return renderApp(coreStart, depsStart as SecurityPluginStartDependencies, params, config);
+          return renderApp(
+            coreStart,
+            depsStart as SecurityPluginStartDependencies,
+            params,
+            config,
+            params.setHeaderActionMenu,
```
Comment (DarshitChanpura) [inline public/plugin.ts:120]: we are already passing params object here. Re-use it?

File: server/plugin.ts [hunk 7]
```diff
@@ -97,6 +105,10 @@ export class SecurityPlugin implements Plugin<SecurityPluginSetup, SecurityPlugi
         plugins: [opensearchSecurityConfigurationPlugin, opensearchSecurityPlugin],
       }
     );
+    if (dataSourceEnabled) {
```
Comment (DarshitChanpura) [inline server/plugin.ts:108]: can you add a brief comment here to help understand what does this block do?

File: server/plugin.ts [hunk 8]
```diff
@@ -133,7 +145,7 @@ export class SecurityPlugin implements Plugin<SecurityPluginSetup, SecurityPlugi
     }
 
     // Register server side APIs
-    defineRoutes(router);
+    defineRoutes(router, dataSourceEnabled);
```
Comment (DarshitChanpura) [inline server/plugin.ts:148]: do you think we can create and utilize a setting `this.config.datasource.enabled` throughout instead of passing `dataSourceEnabled` ?

File: server/routes/index.ts [hunk 9]
```diff
@@ -745,20 +745,39 @@ export function defineRoutes(router: IRouter) {
...
-      try {
-        esResponse = await client.callAsCurrentUser('opensearch_security.clearCache');
-        return response.ok({
-          body: {
-            message: esResponse.message,
-          },
-        });
-      } catch (error) {
-        return errorResponse(response, error);
+      if (!dataSourceEnabled || !request.body?.dataSourceId) {
```
Comment (DarshitChanpura) [inline server/routes/index.ts:755]: if dataSource is enabled and no datasource id is provided, should it still go in this block?

File: server/routes/index.ts [hunk 10]
```diff
@@ -745,20 +745,39 @@ export function defineRoutes(router: IRouter) {
...
+          return response.ok({
+            body: {
+              message: esResponse.message,
+            },
+          });
+        } catch (error) {
+          return errorResponse(response, error);
+        }
+      } else {
+        const client = context.dataSource.opensearch.legacy.getClient(request.body?.dataSourceId);
```
Comment (DarshitChanpura) [inline server/routes/index.ts:769]: this client will be setup by opensearch-core? how do we test this? Where can we set the boundary to say that this is out of scope of testing in security-dashboards-plugin?

File: server/routes/index.ts [hunk 11]
```diff
@@ -745,20 +745,39 @@ export function defineRoutes(router: IRouter) {
...
+        try {
+          esResponse = await client.callAPI('opensearch_security.clearCache', {});
+          return response.ok({
+            body: {
+              message: esResponse.message,
+            },
+          });
+        } catch (error) {
+          return errorResponse(response, error);
+        }
```
Comment (DarshitChanpura) [inline server/routes/index.ts:780]: can we extract this block as a separate function since its duplicate of L757-L767?

File: test/cypress/support/commands.js [hunk 12]
```diff
@@ -91,6 +91,23 @@ Cypress.Commands.add('loginWithSamlMultiauth', () => {
...
+      newOptions = {
+        auth: ADMIN_AUTH,
+      };
+    }
+
+    orig(url, newOptions);
+  } else {
+    orig(url, options);
+  }
+});
```
Comment (DarshitChanpura) [inline test/cypress/support/commands.js:109]:
```suggestion
Cypress.Commands.overwrite('visit', (orig, url, options = {}) => {
  if (Cypress.env('BYPASS_LOGIN')) {
    options.auth = ADMIN_AUTH;
  }
  orig(url, options);
});
```

**Review by DarshitChanpura — COMMENTED** [review]
looks close to me from the first pass. left some comments. Thanks @derek-ho for taking this on!

File: public/plugin.ts (hunk 6 above)
Comment (derek-ho) [reply public/plugin.ts:120]: Sure, good catch

File: public/apps/configuration/top-nav-menu.tsx (hunk 5 above)
Comment (derek-ho) [reply public/apps/configuration/top-nav-menu.tsx:29]: Do you mean the file or the component itself? I think the name SecurityPluginTopNavMenu is fine since that is what it is.

File: public/apps/configuration/panels/get-started.tsx (hunk 3 above)
Comment (derek-ho) [reply public/apps/configuration/panels/get-started.tsx:163]: Renamed to getClusterInfoIfEnabled

File: .github/workflows/cypress-test-multidatasources-enabled-e2e.yml (hunk 1 above)
Comment (derek-ho) [reply .github/workflows/cypress-test-multidatasources-enabled-e2e.yml:34]: See my comment here - https://github.com/opensearch-project/security-dashboards-plugin/pull/1818#discussion_r1525487369, current limitation coming from core, will refactor this after fix becomes available, but it should work for now for our purposes.

File: server/routes/index.ts (hunk 9 above)
Comment (BionIT) [reply server/routes/index.ts:755]: Yes, if data source is enabled, there are 2 situations: 1. user makes a request to local cluster and local cluster doesn't have data source id since it is not a remote data source, in this case, we should still use the existing client; 2. user makes a request to remote data source, then we need to use the data source plugin from core provided client using the data source id. Hope it clarifies!

File: server/plugin.ts (hunk 7 above)
Comment (BionIT) [reply server/plugin.ts:108]: Similar to https://github.com/opensearch-project/security-dashboards-plugin/pull/1818/files#diff-006ee2c1d9e1d5af281d223734f25cceadca71a31ce838157231f2dbbdf63addR105, in order to use the client provided by data source plugin, we need to add the config so that the client knows how to talk to plugin backend by registering the schema. The registerCustomApiSchema essentially does the same for the data source client as in line 105 above

File: server/plugin.ts (hunk 8 above)
Comment (derek-ho) [reply server/plugin.ts:148]: This function is registering some routes on the router, and doesn't have the context of the plugin inside the function. We need to pass it in somehow, I think this is fine for now.

File: server/routes/index.ts (hunk 10 above)
Comment (BionIT) [reply server/routes/index.ts:769]: In this case, we need to add unit test here to test the business logic, and when condition is met, this getClient method is called. In addition, we can add cypress test to test end to end.

File: public/apps/configuration/panels/get-started.tsx (hunk 4 above)
Comment (derek-ho) [reply public/apps/configuration/panels/get-started.tsx:172]: Since the component that consumes this function lives in core, I found it hard to write a unit test for it. I believe the cypress test case that I wrote sufficiently covers this, since it is testing that the message in the toast (which relies on the dataSource set by the setDataSource function) is correctly being displayed after selecting the cluster in the dropdown. Do you think that's enough?

File: test/cypress/support/commands.js (hunk 12 above)
Comment (derek-ho) [reply test/cypress/support/commands.js]: Done

File: server/routes/index.ts (hunk 11 above)
Comment (derek-ho) [reply server/routes/index.ts]: Done

File: server/routes/index.ts (hunk 10 above)
Comment (derek-ho) [reply server/routes/index.ts]: Added tests in: [88fa112](https://github.com/opensearch-project/security-dashboards-plugin/pull/1818/commits/88fa11278e4b6ef938390e9c25ed2de02665cc33), this can be simplified once the TLS issue in core is resolved via: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6171, but it should work for now.

File: .github/workflows/cypress-test-multidatasources-enabled-e2e.yml [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+            - 'CN=node2.dns.a-record,OU=UNIT,O=ORG,L=TORONTO,ST=ONTARIO,C=CA'
+          plugins.security.audit.type: internal_opensearch
+          plugins.security.enable_snapshot_restore_privilege: true
+          plugins.security.check_snapshot_restore_write_privileges: true
+          # TODO: change this back to true/just append to the created opensearch.yml the new port
+          # after the self-signed certs issue is fixed
+          plugins.security.ssl.http.enabled: false 
+          plugins.security.restapi.roles_enabled: [all_access, security_rest_api_access]
+          plugins.security.system_indices.enabled: true
+          plugins.security.system_indices.indices: [.plugins-ml-config, .plugins-ml-connector,
```
Comment (cwperks) [inline .github/workflows/cypress-test-multidatasources-enabled-e2e.yml:61]: Is there any way we can start with the demo security config as a baseline and then make modifications to it like setting the port as 9202 and disabling https (for now)?

This list is subject to change and ideally should be defined in a single place.

File: .github/workflows/cypress-test-multidatasources-disabled-e2e.yml [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+          opensearch_security.cookie.secure: false
+          data_source.enabled: false
+          home.disableWelcomeScreen: true
+          EOT
+
+      - name: Run Cypress Tests
+        uses: ./.github/actions/run-cypress-tests
+        with:
+          dashboards_config_file: opensearch_dashboards_multidatasources.yml
+          yarn_command: 'yarn cypress:run --browser chrome --headless --env BYPASS_LOGIN=true --spec "test/cypress/e2e/multi-datasources/multi_datasources_disabled.spec.js"'
```
Comment (cwperks) [inline .github/workflows/cypress-test-multidatasources-disabled-e2e.yml:49]: Can we rename `BYPASS_LOGIN` to `LOGIN_AS_ADMIN`?

File: public/apps/configuration/configuration-app.tsx [hunk 15]
```diff
@@ -21,14 +21,22 @@ import { I18nProvider } from '@osd/i18n/react';
 import { AppMountParameters, CoreStart } from '../../../../../src/core/public';
 import { SecurityPluginStartDependencies, ClientConfigType } from '../../types';
 import { AppRouter } from './app-router';
+import { DataSourceManagementPluginSetup } from '../../../../../src/plugins/data_source_management/public';
 
 export function renderApp(
   coreStart: CoreStart,
-  navigation: SecurityPluginStartDependencies,
+  securityPluginStartDeps: SecurityPluginStartDependencies,
```
Comment (cwperks) [inline public/apps/configuration/configuration-app.tsx:28]: nit: can we rename this to depsStart. I see this naming used across OSD core in a few plugins.

Can we also change every instance of where `navigation` appears to `depsStart`?

File: public/apps/configuration/panels/test/get-started.test.tsx [hunk 16]
```diff
@@ -81,6 +83,7 @@ describe('Get started (landing page)', () => {
           navigation={{} as any}
           params={{} as any}
           config={config as any}
+          securityPluginStartDeps={{}}
```
Comment (cwperks) [inline public/apps/configuration/panels/test/get-started.test.tsx:86]: This variable replaces `navigation`. Can you change the test to reflect the change in name from `navigation` -> `depsStart` or `securityPluginStartDeps`

File: server/routes/index.ts [hunk 17]
```diff
@@ -745,21 +747,20 @@ export function defineRoutes(router: IRouter) {
   router.delete(
     {
       path: `${API_PREFIX}/configuration/cache`,
-      validate: false,
+      validate: {
+        body: schema.object({
```
Comment (cwperks) [inline server/routes/index.ts:751]: Should the body of this request be validated if data sources are disabled?

Would datasourceId be applicable to all endpoints?

Comment (cwperks) [reply server/routes/index.ts:751]: Can we get the current datasource id from local storage and choose the correct client accordingly in the API handler and not require dataSourceId in the body of this request?

File: server/routes/index.ts [hunk 18]
```diff
@@ -883,6 +884,42 @@ export function defineRoutes(router: IRouter) {
...
+        },
+      });
+    } catch (error) {
+      return errorResponse(response, error);
+    }
+  } else {
+    const client = context.dataSource.opensearch.legacy.getClient(request.body?.dataSourceId);
+    let esResponse;
+    try {
+      esResponse = await client.callAPI(endpoint, {});
```
Comment (cwperks) [inline server/routes/index.ts:911]: Can this be renamed as `callAsCurrentUser` so that it matches the block above? Ideally, the only thing that is different is the http client between the 2 blocks.

File: .github/workflows/cypress-test-multidatasources-enabled-e2e.yml (hunk 13 above)
Comment (derek-ho) [reply .github/workflows/cypress-test-multidatasources-enabled-e2e.yml:61]: Yes - I did so in: https://github.com/derek-ho/start-opensearch/blob/v3/action.yml#L132 for OpenSearch's running without security, just by adding a port number that is not standard. However the issue is that the security demo configuration by default writes https enabled, and you cannot simply add on a disable block of the same setting, since it would complain about duplicate fields. Thus as a workaround, this is a different branch where I instead replace the whole opensearch.yml. Once your fix goes into core, we can re-enable https and simply add a port number to the opensearch.yml after the security demo configuration, instead of rewriting the whole thing. Going to mark this as resolved for now, but feel free to re-open if that didn't cover your concern.

File: public/apps/configuration/configuration-app.tsx (hunk 15 above)
Comment (derek-ho) [reply public/apps/configuration/configuration-app.tsx:28]: Yes - I thought navigation was misleading and named it to securityPluginStartDeps and changed all the references. I can change it to depsStart, to follow the convention.

File: public/apps/configuration/panels/test/get-started.test.tsx (hunk 16 above)
Comment (derek-ho) [reply public/apps/configuration/panels/test/get-started.test.tsx:86]: We don't really use any of the deps in other tests, other than the new ones that I added to verify the datasourceEnabled functionality - do you have any sugestions on testing the change/rename, other than the existing tests passing?

File: server/routes/index.ts (hunk 17 above)
Comment (cwperks) [reply server/routes/index.ts:751]: I'm looking into seeing if its possible to use `sessionStorageFactory` to keep track of the current data source and use what's stored to determine the client to you to avoid having to pass datasourceId into all of the routes.

File: public/apps/configuration/panels/test/get-started.test.tsx (hunk 16 above)
Comment (cwperks) [reply public/apps/configuration/panels/test/get-started.test.tsx:86]: The existing tests passing will be fine since its a refactor of a variable and no functionality is changed.

I left a comment here because `navigation` and `securityPluginStartDeps` are both being passed in as props, but this PR shows that `navigation` is being renamed to `securityPluginStartDeps`, so this should be modified to remove references to the old name `navigation`.

Comment (derek-ho) [reply public/apps/configuration/panels/test/get-started.test.tsx:86]: Oh ok, I misunderstood what you meant, sure removing the old navigation variable makes sense to me

File: server/routes/index.ts (hunk 17 above)
Comment (derek-ho) [reply server/routes/index.ts:751]: Repurposed this issue: https://github.com/opensearch-project/security-dashboards-plugin/issues/1816, to decide on how to maintain the context within our plugin. I will not address this question in this PR and will track it there. Right now the two competing ways of doing so would be to lift the state/setstate one level up to be shared by all tabs, or by using session/local storage so each request wrapped with the datasource helper function can read from it instead of getting passed one by the individual tabs.

File: server/routes/index.ts (hunk 18 above)
Comment (derek-ho) [reply server/routes/index.ts:911]: This is the equivalent to callAsCurrentUser from the legacy client if I am understanding correctly. Used this to stay consistent with the other parts of the plugin where we are using legacy clients. https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/docs/multi-datasource/client_management_design.md#421-legacy-client

**Review by cwperks — APPROVED** [review]

File: server/routes/index.ts (hunk 17 above)
Comment (cwperks) [inline server/routes/index.ts:751]: Should dataSourceId(s) always be passed in as query params?

Comment (derek-ho) [reply server/routes/index.ts:751]: Yea, since we control all the places this is called (only in the get started tab) I think it's fine to require it. This isn't using query params though it's only an internal implementation. However I think this is a two way door decision and once more routes are modified we can easily refactor it to a cleaner way if we find one

Comment (derek-ho) [reply server/routes/index.ts:751]: If that makes sense can you merge? Thanks!

**Review by DarshitChanpura — APPROVED** [review]
